File: script/tec3D.py
#!/bin/python3

from pathlib import Path
import math
import numpy as np
import re, os
import sys
from pyiner import fini_get
from mesh_refine import refine_by_splitting

print(f"  Tecplot 3D mesh generation")
print()

ini_file = 'input.ini'
if not Path(ini_file).is_file():
    print('  ERROR : INI file not found -> mesh.tec generation interrupted')
    exit()

circular_sector = False
threedim = False

outpath = fini_get(ini_file, 'PATCHWORK', 'outpath')
if outpath:
    fileoutput = outpath + "/mesh.tec"

fs = fini_get(ini_file, 'PATCHWORK', 'fs', float)
if not fs:
    fs = 1.0

includefile = fini_get(ini_file, 'PATCHWORK', 'includefile')
if includefile:
    ini_file = includefile

fileinput = fini_get(ini_file, 'PATCHWORK', 'inputfile')
if fileinput:
    expanded_path = os.path.expanduser(fileinput)
    folder_path = os.path.dirname(expanded_path)
    if len(folder_path) == 0:
        folder_path = "."
    method = fini_get(ini_file, 'PATCHWORK', 'method')
    if method:
        fileinput = folder_path + "/mesh.dat"
        fileoutput = folder_path + "/mesh.tec"
    else:
        fileoutput = folder_path + "/mesh.tec"
elif outpath:
    fileinput = outpath + "/mesh.dat"
    fileoutput = outpath + "/mesh.tec"
else:
    fileinput = 'mesh.dat'
    fileoutput = 'mesh.tec'

# Read mesh type and input (2D,2Daxi,3D)
type = fini_get(ini_file,'PATCHWORK','type')
if type:
    if type=='2D':
        Nz_fix = 2
        width_max = fini_get(ini_file, 'PATCHWORK', 'width', float)
        if not width_max:
            width_max = 1e-3
        zmax = width_max
        stretch = 'null'
    elif type == '2Daxi':
        Nz_fix = 2
        theta_max = fini_get(ini_file, 'PATCHWORK', 'theta', float)
        if not theta_max:
            theta_max = 1*np.pi/180
        zmax = theta_max
        stretch = 'null'
        circular_sector = True
    elif type == '3D':
        Nz_fix = fini_get(ini_file, 'PATCHWORK', 'Nz')
        if Nz_fix:
            Nz_fix = int(Nz_fix)+1
            theta_max = fini_get(ini_file, 'PATCHWORK', 'theta', float)
            width_max = fini_get(ini_file, 'PATCHWORK', 'width', float)
            if theta_max:
                theta_max = theta_max*np.pi/180
                zmax = theta_max
                circular_sector = True
            elif width_max:
                zmax = width_max
            else:
                exit('  ERROR : theta or width parameter necessary!')
            stretch = fini_get(ini_file, 'PATCHWORK', 'stretch')
            if stretch:
                c = fini_get(ini_file, 'PATCHWORK', 'c')
                p = fini_get(ini_file, 'PATCHWORK', 'p')
            else:
                stretch = 'null'
        else:
            threedim = True
            print(f"  Look for a 3D mesh file")
else:
    stretch = 'null'
    threedim = True
    print(f"  Look for a 3D mesh file")

axis = fini_get(ini_file,'PATCHWORK','axis', float)

if not threedim:
    a = np.linspace(0,Nz_fix,num=Nz_fix)/Nz_fix
    if stretch == 'null':
        # No stretch
        zdir = zmax*a-0.5*zmax
    elif stretch == 'left':
        # Left
        zdir = zmax*np.tanh(c*a)/np.tanh(c)-0.5*zmax
    elif stretch == 'outer':
        # Outer
        zdir = zmax*(np.tanh(c*(a-p))+np.tanh(c*p))/(np.tanh(c*(1-p))+np.tanh(c*p))-0.5*zmax

if threedim:
    # Define the source and destination file names
    content_to_remove = "DT=(SINGLE SINGLE)"
    # Open the source file for readin
    with open(fileinput, "r") as source_file:
        # Open the destination file for writing
        with open(fileoutput, "w") as destination_file:
            # Iterate through each line in the source file
            for line in source_file:
                # Check if the line contains the content to remove
                if content_to_remove not in line:
                    # If it doesn't contain the content, write it to the destination file
                    destination_file.write(line)
    print("  Done!")
    exit()

# Read mesh.dat 2D
MESH = open(fileinput,"r")
MESH = MESH.readlines()

Nb = 0

Nx = []
Ny = []
Nz = []

xb = []
yb = []
zb = []

line = 3

while line <= len(MESH)-1:
    
    Nb   += 1
    line += 1
        
    # Extract Nx, Ny, and Nz

    # Define the regular expression pattern to match numbers
    pattern = r'\b\d+\b'
    numbers = re.findall(pattern, MESH[line])
    # Convert the extracted strings to integers
    numbers = [int(num) for num in numbers]

    Nx.append(numbers[0])
    Ny.append(numbers[1])
    Nz.append(numbers[2])
    
    # Skip useless lines
    line += 3

    # Extract X and Y coordinates

    x2d = []
    for j in range(Ny[Nb-1]):
        x1d = []
        for i in range(Nx[Nb-1]):
            x = float(MESH[line])
            x1d.append(x)
            line += 1
        x2d.append(x1d)
    
    xb.append(x2d)
    
    y2d = []
    for j in range(Ny[Nb-1]):
        y1d = []
        for i in range(Nx[Nb-1]):
            y = float(MESH[line])
            if axis and y <= axis:
                y = axis
            y1d.append(y)
            line += 1
        y2d.append(y1d)
    
    yb.append(y2d)

if not threedim:
    Nz[:] = [Nz_fix]*len(Nz)

# Build 3D mesh
print(f"  Building up the 3D grid ...")

Xf = []; Yf = []; Zf = []
XB = []; YB = []; ZB = []
for b in range(Nb):
    X3D = []
    Y3D = []
    Z3D = []
    for k in range(Nz[b]):
        zz = zdir[k]
        X2D = []
        Y2D = []
        Z2D = []
        for j in range(Ny[b]):
            X1D = []
            Y1D = []
            Z1D = []
            for i in range(Nx[b]):
                X1D.append(xb[b][j][i])
                if circular_sector:
                    Y1D.append(yb[b][j][i]*math.cos(zz))
                    Z1D.append(yb[b][j][i]*math.sin(zz))
                else:
                    Y1D.append(yb[b][j][i])
                    Z1D.append(zz)
            X2D.append(X1D)
            Y2D.append(Y1D)
            Z2D.append(Z1D)        
        X3D.append(X2D)
        Y3D.append(Y2D)
        Z3D.append(Z2D)
        zz += zmax/(Nz[b]-1)
    
    # Check for refinement request
    refine = fini_get(ini_file, 'PATCHWORK', 'refine', bool)
    if not refine:
        refine = False
    if refine:
        [Xf,Yf,Zf] = refine_by_splitting(X3D,Y3D,Z3D)
        XB.append(Xf)
        YB.append(Yf)
        ZB.append(Zf)
        Nx[b] = len(Xf[0][0])
        Ny[b] = len(Xf[0])
        Nz[b] = len(Xf)
    else:
        XB.append(X3D)
        YB.append(Y3D)
        ZB.append(Z3D)
    

# Write mesh.tec
MESH = open(fileoutput,"w")
MESH.writelines(' TITLE     = "PATCHWORK mesh" \n')
MESH.writelines(' VARIABLES = "X", "Y", "Z" \n')

for b in range(Nb):
    MESH.writelines(' ZONE T="BLOCCO '+"{0:>1}".format(str(b+1))+'" \n')
    MESH.writelines(' I='+"{0:>3}".format(str(Nx[b]))+", J="+"{0:>3}".format(str(Ny[b]))+", K="+"{0:>3}".format(str(Nz[b]))+", ZONETYPE=Ordered \n")
    MESH.writelines(' DATAPACKING=BLOCK \n')
    # MESH.writelines(' DT=(SINGLE SINGLE) \n')

    for k in range(Nz[b]):
        for j in range(Ny[b]):
            for i in range(Nx[b]):
                MESH.writelines(str(XB[b][k][j][i]/fs)+"\n")

    for k in range(Nz[b]):
        for j in range(Ny[b]):
            for i in range(Nx[b]):
                MESH.writelines(str(YB[b][k][j][i]/fs)+"\n")

    for k in range(Nz[b]):
        for j in range(Ny[b]):
            for i in range(Nx[b]):
                MESH.writelines(str(ZB[b][k][j][i]/fs)+"\n")

print(f"  Done!")

File: src/tec3D.f90
module Tec3D_mod
    use finer, only : file_ini
    use iso_fortran_env, only: I4 => int32, R8 => real64
    use Data_Types
    use linspace_mod
    implicit none
    contains
  
    subroutine Tec3D (grid)
      implicit none
      type(Grid_Type), intent(in) :: grid
      integer :: Nb, Nx, Ny, Nz, i, j, k, b, Nz_fix, error
      integer, allocatable :: Nx_arr(:), Ny_arr(:), Nz_arr(:)
      real(R8), allocatable :: X(:,:,:,:), Y(:,:,:,:), Z(:,:,:,:)
      real(R8), allocatable :: a(:), zdir(:)
      real(R8) :: axis, zmax, c, p, theta_max, width_max
      character(len=256) :: inputfile, outputfile, mesh_type, stretch
      logical :: threedim, circular_sector
      type(file_ini) :: fini
  
      inputfile = 'mesh.dat'
      outputfile = 'mesh.tec'
      circular_sector = .false.
      threedim = .false.
      axis = 0_R8

      call fini%load(filename='input.ini')
      call fini%get(section_name='PATCHWORK', option_name='inputfile', val=inputfile, error=error)
      call fini%get(section_name='PATCHWORK', option_name='outputfile', val=outputfile, error=error)
      call fini%get(section_name='PATCHWORK', option_name='type', val=mesh_type, error=error)
      if (error/=0) mesh_type = '3D'
      call fini%get(section_name='PATCHWORK', option_name='axis', val=axis, error=error)

      select case(mesh_type)
      case('2D')
        Nz_fix = 2
        call fini%get(section_name='PATCHWORK', option_name='width', val=width_max, error=error)
        if (error/=0) then
          width_max = 1e-3
          zmax = width_max
          stretch = 'none'
        end if
      case('2Daxi')
        Nz_fix = 2
        call fini%get(section_name='PATCHWORK', option_name='theta', val=theta_max, error=error)
        if (error/=0) then
          theta_max = 3.141592653589793_R8 / 180_R8
          zmax = theta_max
          circular_sector = .true.
        end if
        stretch = 'none'
      case('3D')
        call fini%get(section_name='PATCHWORK', option_name='Nz', val=Nz_fix, error=error)
        if (error==0) then
          Nz_fix = Nz_fix + 1
          call fini%get(section_name='PATCHWORK', option_name='theta', val=theta_max, error=error)
          if (error==0) then
            theta_max = theta_max * 3.141592653589793_R8 / 180_R8
            zmax = theta_max
            circular_sector = .true.
          else
            call fini%get(section_name='PATCHWORK', option_name='width', val=width_max, error=error)
            if (error/=0) error stop ( '  ERROR : theta or width parameter necessary!' )
            zmax = width_max
          end if
          call fini%get(section_name='PATCHWORK', option_name='stretch', val=stretch, error=error)
          if (error==0) then
            call fini%get(section_name='PATCHWORK', option_name='c', val=c, error=error)
            call fini%get(section_name='PATCHWORK', option_name='p', val=p, error=error)
          else
            stretch = 'none'
          end if
        else
          threedim = .true.
          write(*,*) '  Look for a 3D mesh file'
        end if
      end select

      if (.not.threedim) then
        a = linspace(0d0,real(Nz_fix,kind=8),Nz_fix)/Nz_fix
        if (trim(stretch) == 'none') then
          ! No stretch
          zdir = zmax*a-0.5_R8*zmax
          print*, 'no stretch ok'
        elseif (trim(stretch) == 'left') then
          ! Left
          zdir = zmax*tanh(c*a)/tanh(c)-0.5*zmax
        elseif (trim(stretch) == 'outer') then
          ! Outer
          zdir = zmax*(tanh(c*(a-p))+tanh(c*p))/(tanh(c*(1-p))+tanh(c*p))-0.5*zmax
        end if
      end if

      ! 3D mesh
      Nb = grid%nblocks
     
      allocate(Nx_arr(Nb), Ny_arr(Nb), Nz_arr(Nb))
     
      ! Read block dimensions
      do b = 1, Nb
        Nx_arr(b) = grid%blk(b)%ni+1
        Ny_arr(b) = grid%blk(b)%nj+1
        Nz_arr(b) = Nz_fix
      end do
     
      ! Find max dimensions to allocate arrays
      Nx = maxval(Nx_arr)
      Ny = maxval(Ny_arr)
      Nz = maxval(Nz_arr)
     
      allocate(X(Nx,Ny,Nz,Nb), Y(Nx,Ny,Nz,Nb), Z(Nx,Ny,Nz,Nb))
     
      ! Read mesh data
      do b = 1, Nb
        do j = 1, Ny_arr(b)
          do i = 1, Nx_arr(b)
            X(i,j,1,b) = grid%blk(b)%x(i,j)
          end do
        end do
        do j = 1, Ny_arr(b)
          do i = 1, Nx_arr(b)
            Y(i,j,1,b) = grid%blk(b)%y(i,j)
            if (Y(i,j,1,b) <= axis) Y(i,j,1,b) = axis
          end do
        end do
      end do
     
      ! Handle circular sector transformation
      if (circular_sector) then
        do b = 1, Nb
          do j = 1, Ny_arr(b)
            do i = 1, Nx_arr(b)
              Z(i,j,1,b) = Y(i,j,1,b) !* sin(theta_max)
              Y(i,j,1,b) = Y(i,j,1,b) !* cos(theta_max)
            end do
          end do
        end do
      end if
     
      ! ! Stretching in z-direction
      ! do b = 1, Nb
      !   do k = 1, Nz_arr(b)
      !     Z(:,:,k,b) = (real(k-1)/real(Nz_arr(b)-1)) * zmax - 0.5_R8 * zmax
      !   end do
      ! end do
     
      ! Open output file
      open(20, file=outputfile, status='unknown', action='write')
      write(20, '(A)') 'TITLE = "PATCHWORK mesh"'
      write(20, '(A)') 'VARIABLES = "X", "Y", "Z"'
     
      ! Write mesh data
      do b = 1, Nb
        write(20, '(A,I4,A,I4,A,I4,A)') 'ZONE I=', Nx_arr(b), ', J=', Ny_arr(b), ', K=', Nz_arr(b), ', ZONETYPE=Ordered'
        write(20, '(A)') 'DATAPACKING=BLOCK'
       
        do k = 1, Nz_arr(b)
          do j = 1, Ny_arr(b)
            do i = 1, Nx_arr(b)
              write(20, '(F12.6)') X(i,j,k,b)
            end do
          end do
        end do
       
        do k = 1, Nz_arr(b)
          do j = 1, Ny_arr(b)
            do i = 1, Nx_arr(b)
              write(20, '(F12.6)') Y(i,j,k,b)
            end do
          end do
        end do
       
        do k = 1, Nz_arr(b)
          do j = 1, Ny_arr(b)
            do i = 1, Nx_arr(b)
              write(20, '(F12.6)') Z(i,j,k,b)
            end do
          end do
        end do
      end do
      close(20)
     
      ! Cleanup
      deallocate(X, Y, Z, Nx_arr, Ny_arr, Nz_arr)
     
      print *, 'Mesh processing complete.'
    end subroutine Tec3D
  
  end module Tec3D_mod